File: src/309_best_time_to_buy_and_sell_stock_with_cooldown/best_time_to_buy_and_sell_stock_with_cooldown.py
from typing import List


class Solution:
    def maxProfitTopDown(self, prices: List[int]) -> int:
        hold = float("-inf")
        sold = 0
        colldown = 0

        for price in prices:
            prev_sold = sold

            sold = hold + price  # sell today
            hold = max(hold, colldown - price)  # buy or keep holding
            colldown = max(colldown, prev_sold)  # cooldown or stay resting

        return max(sold, colldown)

    def maxProfitTopDownTopDown(self, prices: List[int]) -> int:
        memo = {}

        def dfs(i, buy):
            if (i, buy) in memo:
                return memo[(i, buy)]
            if i >= len(prices):
                return 0

            cooldown = dfs(i + 1, buy)
            res = 0
            if buy:
                res = max(dfs(i + 1, False) - prices[i], cooldown)
            else:
                res = max(dfs(i + 2, True) + prices[i], cooldown)

            memo[(i, buy)] = res
            return res

        return dfs(0, True)


# Helper functions for testing


def test_case(prices, expected):
    sol = Solution()
    actual = sol.maxProfit(prices)
    print(f"{actual} | {expected}")


# Test cases

if __name__ == "__main__":
    test_case([1, 2, 3, 0, 2], 3)
    test_case([1], 0)
    test_case([1, 2], 1)
    test_case([2, 1], 0)
    test_case([1, 2, 4], 3)
    test_case([1, 3, 2, 4], 4)

File: src/309_best_time_to_buy_and_sell_stock_with_cooldown/BestTimeToBuyAndSellStockWithCooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        long hold = Long.MIN_VALUE;
        int sold = 0;
        int rest = 0;

        for (int price : prices) {
            int prevSold = sold;
            sold = (int) (hold + price);
            hold = Math.max(hold, rest - price);
            rest = Math.max(rest, prevSold);
        }

        return Math.max(sold, rest);
    }
}

class BestTimeToBuyAndSellStockWithCooldown {
    public static void main(String[] args) {
        Solution sol = new Solution();

        // Test cases
        testCase(sol, new int[] { 1, 2, 3, 0, 2 }, 3);
        testCase(sol, new int[] { 1 }, 0);
        testCase(sol, new int[] { 1, 2 }, 1);
        testCase(sol, new int[] { 2, 1 }, 0);
        testCase(sol, new int[] { 1, 2, 4 }, 3);
        testCase(sol, new int[] { 1, 3, 2, 4 }, 4);
    }

    private static void testCase(Solution sol, int[] prices, int expected) {
        int actual = sol.maxProfit(prices);
        System.out.println(actual + " | " + expected);
    }
}
